### PR TITLE
Implement `uv init`

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

Implements the `uv init` command, which initializes a project (`pyproject.toml`, `README.md`, `src/__init__.py`) in the current directory, or in the given path. `uv init` also does workspace discovery.

Resolves https://github.com/astral-sh/uv/issues/1360.

### CODE REVIEW

File: crates/uv/src/commands/project/init.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    // Create the `pyproject.toml`.
+    fs_err::write(
+        canonical_path.join("pyproject.toml"),
+        indoc::formatdoc! {r#"
+        [project]
+        name = "{name}"
+        version = "0.1.0"
+        description = "Add your description here"
+        dependencies = []
+        readme = "README.md"
```
Comment (ibraheemdev) [inline crates/uv/src/commands/project/init.rs:84]: Anything else we should add here? I think Rye and Poetry also fill in the authors field, probably through git config?

Comment (ibraheemdev) [reply crates/uv/src/commands/project/init.rs:88]: `uv init` also doesn't currently run `git init` or support `--vcs`, which cargo does.

File: crates/uv-cli/src/lib.rs [hunk 2]
```diff
@@ -1641,6 +1644,42 @@ impl ExternalCommand {
...
+    #[arg(long)]
+    pub name: Option<PackageName>,
+
+    /// Do not create a readme file.
+    #[arg(long)]
+    pub no_readme: bool,
+
+    /// Do not create a `.python-version` file.
+    #[arg(long)]
+    pub no_pin: bool,
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:1666]: I'm not sure we want to create this file. We need to support pinning via the `pyproject.toml` too.

File: crates/uv/src/commands/project/init.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    EnvironmentPreference, PythonFetch, PythonInstallation, PythonPreference, PythonRequest,
+};
+use uv_warnings::warn_user_once;
+
+use crate::commands::ExitStatus;
+use crate::printer::Printer;
+
+/// Add one or more packages to the project requirements.
+#[allow(
+    clippy::too_many_arguments,
```
Comment (zanieb) [inline crates/uv/src/commands/project/init.rs:23]: I thought we dropped this lint

File: crates/uv/src/commands/project/init.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    let current_dir = std::env::current_dir()?.canonicalize()?;
+    let workspace = match ProjectWorkspace::discover(&current_dir, None).await {
+        Ok(project) => Some(project),
+        Err(WorkspaceError::MissingPyprojectToml) => None,
+        Err(err) => return Err(err.into()),
+    };
+
+    let canonical_path = match path {
+        None => current_dir.clone(),
+        Some(ref path) => current_dir.join(path),
```
Comment (zanieb) [inline crates/uv/src/commands/project/init.rs:55]: What if we're given an absolute path?

Comment (ibraheemdev) [reply crates/uv/src/commands/project/init.rs]: Fixed.

**Comment by zanieb** [comment]
Sorry I only got part way through reviewing and had to go — I'll try to give it another look soon.

File: crates/uv/src/commands/project/init.rs (hunk 1 above)
Comment (imbev) [reply crates/uv/src/commands/project/init.rs:85]: > Anything else we should add here? I think Rye and Poetry also fill in the authors field, probably through git config?

The command could prompt, using the git config or example, example@example.com as a default.

Comment (imbev) [reply crates/uv/src/commands/project/init.rs:85]: I'm not sure if it's within the scope of this PR, but there could be a boolean prompt that could lead to further prompting to add package-specific fields such as keywords, classifiers, urls

File: crates/uv/src/commands/project/init.rs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    // Create the `pyproject.toml`.
+    fs_err::write(
+        project_dir.join("pyproject.toml"),
+        indoc::formatdoc! {r#"
+        [project]
+        name = "{name}"
+        version = "0.1.0"
+        description = "Add your description here"
+        dependencies = []
+        readme = "README.md"
```
Comment (bluss) [inline crates/uv/src/commands/project/init.rs:85]: Ideally only fill in the readme key if actually creating the README. (Rye has a nice goal that it creates a working configuration without extra steps.)

**Comment by zanieb** [comment]
Let me know when this is ready for another look

File: crates/uv/src/commands/project/init.rs (hunk 1 above)
Comment (ibraheemdev) [reply crates/uv/src/commands/project/init.rs]: The other thing missing is `requires-python`, should we default to `>=3.12`?

Comment (imbev) [reply crates/uv/src/commands/project/init.rs]: uv has a default python executable, perhaps requires-python could be set to that?

Comment (zanieb) [reply crates/uv/src/commands/project/init.rs]: Defaulting to the latest seems.. reasonable.

**Review by zanieb — APPROVED** [review]
Let's go :)

File: crates/uv-cli/src/lib.rs (hunk 2 above)
Comment (zanieb) [reply crates/uv-cli/src/lib.rs]: ref #4970

**Pull request merged**